src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs

```diff
@@ -110,68 +110,193 @@ public override void Initialize(AnalysisContext analysisContext)
+
+                return IsThrowsArgument(parameterSymbol, "testCode", methodNames, "Xunit.Assert");
+            }
+
+            // We skip analysis for the last statement in a lambda passed to Assert.Throws/ThrowsAsync (xUnit and NUnit), or the last
+            // statement in a method annotated with [ExpectedException] (MSTest)
+
+            // Note: We do not attempt to account for a synchronously-running ThrowsAsync with something like return Task.CompletedTask;
+            // as the last line.
+
+            // We only skip analysis if we're in a method
+            if (operationContext.ContainingSymbol.Kind != SymbolKind.Method)
+            {
+                return false;
+            }
+
+            // Get the enclosing block. If that block's parent isn't null (MSTest case) or an IAnonymousFunctionOperation (xUnit/NUnit), then
+            // we bail immediately
+            if (!(operationContext.Operation.Parent is IBlockOperation enclosingBlock))
+            {
+                return false;
+            }
+
+            if (enclosingBlock.Parent != null && enclosingBlock.Parent.Kind != OperationKind.AnonymousFunction)
+            {
+                return false;
+            }
+
+            // Only analyze the last non-implicit statement in the function
+            bool foundBlock = false;
+            foreach (var statement in enclosingBlock.Operations)
+            {
+                if (statement == operationContext.Operation)
+                {
+                    foundBlock = true;
+                    continue;
+                }
+                else if (foundBlock)
+                {
+                    if (!statement.IsImplicit)
+                    {
+                        return false;
+                    }
+                }
+                else
+                {
+                    continue;
+                }
+            }
+
+            // If the parent is Null, we're in the MSTest case. Otherwise, we're in the xUnit/NUnit case.
+            if (enclosingBlock.Parent == null)
+            {
+                if (expectedExceptionType == null)
+                {
+                    return false;
+                }
+
+                IMethodSymbol methodSymbol = (IMethodSymbol)operationContext.ContainingSymbol;
+
+                return methodSymbol.GetAttributes().Any(attr => attr.AttributeClass == expectedExceptionType);
+            }
+            else
+            {
+                // Look for an enclosing IArgumentOperation
+                IOperation parentArgument = enclosingBlock;
+                do
+                {
+                    parentArgument = parentArgument.Parent;
+                } while (parentArgument != null && parentArgument.Kind != OperationKind.Argument);
+
+                if (parentArgument == null)
+                {
+                    return false;
+                }
+
+                IArgumentOperation argumentOperation = (IArgumentOperation)parentArgument;
```

Comment:
Looks like a good candidate for an extension method `TOperation GetAncestor<TOperation>(this IOperation operation, OperationKind ancestorKind) where TOperation: IOperation` for AnalyzerUtilities?